modules/core/src/main/java/org/apache/ignite/internal/util/nio/compression/GridNioCompressionHandler.java

```diff
@@ -0,0 +1,274 @@
+            inNetBuf.put(encBuf);
+        }
+
+        // Initially buffer is empty.
+        outNetBuf.position(0);
+        outNetBuf.limit(0);
+
+        int appBufSize = NET_BUF_SIZE * 2;
+
+        appBuf = directBuf ? ByteBuffer.allocateDirect(appBufSize) : ByteBuffer.allocate(appBufSize);
+        appBuf.order(order);
+
+        if (log.isDebugEnabled())
+            log.debug("Started compress session [netBufSize=" + NET_BUF_SIZE + ", appBufSize=" + appBufSize + ']');
+    }
+
+    /**
+     * @return Application buffer with decoded data.
+     */
+    ByteBuffer getApplicationBuffer() {
+        return appBuf;
+    }
+
+    /**
+     * Called by compress filter when new message was received.
+     *
+     * @param buf Received message.
+     * @throws GridNioException If exception occurred while forwarding events to underlying filter.
+     * @throws IOException If failed to process compress data.
+     */
+    void messageReceived(ByteBuffer buf) throws IgniteCheckedException, IOException {
+        if (buf.limit() > inNetBuf.remaining()) {
+            assert inNetBuf.capacity() + buf.limit() * 2 <= Integer.MAX_VALUE;
+
+            inNetBuf = expandBuffer(inNetBuf, inNetBuf.capacity() + buf.limit() * 2);
+
+            if (log.isDebugEnabled())
+                log.debug("Expanded buffer [inNetBufCapacity=" + inNetBuf.capacity() + "], ses=" + ses + ", ");
+        }
+
+        // append buf to inNetBuffer
+        inNetBuf.put(buf);
+
+        unwrapData();
+    }
+
+    /**
+     * Compress data to be written to the network.
+     *
+     * @param src data to compress.
+     * @return Output buffer with compressed data.
+     * @throws IOException on errors.
+     */
+    ByteBuffer compress(ByteBuffer src) throws IOException {
+        assert isHeldByCurrentThread();
+
+        // The data buffer is (must be) empty, we can reuse the entire
+        // buffer.
+        outNetBuf.clear();
+
+        // Loop until there is no more data in src
+        while (src.hasRemaining()) {
+            CompressionEngineResult res = compressionEngine.compress(src, outNetBuf);
+
+            if (res == BUFFER_OVERFLOW) {
+                assert outNetBuf.capacity() <= Integer.MAX_VALUE / 2;
+
+                outNetBuf = expandBuffer(outNetBuf,outNetBuf.capacity() * 2);
+
+                if (log.isDebugEnabled())
+                    log.debug("Expanded output net buffer [outNetBufCapacity=" + outNetBuf.capacity() + ", ses=" +
+                        ses + ']');
+            }
+        }
+
+        outNetBuf.flip();
+
+        return outNetBuf;
+    }
+
+    /**
+     * Copies data from out net buffer and passes it to the underlying chain.
+     *
+     * @param ackC Closure invoked when message ACK is received.
+     * @return Write future.
+     * @throws GridNioException If send failed.
+     */
+    GridNioFuture<?> writeNetBuffer(IgniteInClosure<IgniteException> ackC) throws IgniteCheckedException {
```

Comment:
Add `@Nullable` to `ackC`. See `shutdownSession` method which pass `null`.

Reply:
Done.